sklearn/utils/estimator_checks.py

```diff
@@ -1339,6 +1344,31 @@ def check_estimators_dtypes(name, estimator_orig):
                 getattr(estimator, method)(X_train)
 
 
+def check_estimators_preserve_dtypes(name, estimator_orig):
+
+    X = np.random.RandomState(0).randn(50, 5)
+    y = np.random.RandomState(0).randn(50)
+
+    for dtype_in, dtype_out in [(np.float32, np.float32),
+                                (np.float64, np.float64),
+                                (np.float16, np.float64)]:
```

Comment:
StandardScaler will not pass this test because it preserving even 16 bits without casting which is fine.

Basically, for dtype in the `preserves_dtype` list, we should ensure that the dtype is the same, otherwise, it should be casted to 64 bits.